src/sentry/discover/endpoints/serializers.py

```diff
@@ -131,3 +133,90 @@ def get_condition(self, condition):
+
+
+class DiscoverSavedQuerySerializer(serializers.Serializer):
+    name = serializers.CharField(required=True)
+    projects = ListField(child=serializers.IntegerField(), required=False, default=[])
+    start = serializers.DateTimeField(required=False, allow_null=True)
+    end = serializers.DateTimeField(required=False, allow_null=True)
+    range = serializers.CharField(required=False, allow_null=True)
+    fields = ListField(child=serializers.CharField(), required=False, allow_null=True)
+    orderby = serializers.CharField(required=False, allow_null=True)
+
+    # This block of fields is only accepted by discover 1 which omits the version
+    # attribute or has it set to 1
+    rollup = serializers.IntegerField(required=False, allow_null=True)
+    aggregations = ListField(child=ListField(), required=False, allow_null=True)
+    groupby = ListField(child=serializers.CharField(), required=False, allow_null=True)
+    conditions = ListField(child=ListField(), required=False, allow_null=True)
+    limit = serializers.IntegerField(min_value=0, max_value=1000, required=False, allow_null=True)
+
+    # There are multiple versions of saved queries supported.
+    version = serializers.IntegerField(min_value=1, max_value=2, required=False, allow_null=True)
+
+    # Attributes that are only accepted if version = 2
+    environment = ListField(child=serializers.CharField(), required=False, allow_null=True)
+    fieldnames = ListField(child=serializers.CharField(), required=False, allow_null=True)
+    query = serializers.CharField(required=False, allow_null=True)
+
+    def validate_projects(self, projects):
+        organization = self.context["organization"]
+
+        org_projects = set(
+            Project.objects.filter(
+                organization=organization, id__in=projects, status=ProjectStatus.VISIBLE
+            ).values_list("id", flat=True)
+        )
+
+        if set(projects) != org_projects:
+            raise PermissionDenied
+
+        return projects
+
+    def validate(self, data):
+        query = {}
+        query_keys = [
+            "fieldnames",
+            "environment",
+            "query",
+            "version",
+            "fields",
+            "conditions",
+            "aggregations",
+            "range",
+            "start",
+            "end",
+            "orderby",
+            "limit",
+        ]
+
+        for key in query_keys:
+            if data.get(key) is not None:
+                query[key] = data[key]
+
+        if query.get("version", 1) == 1:
+            not_allowed = set(query.keys()) & set(["environment", "fieldnames", "query"])
+            if not_allowed:
+                raise serializers.ValidationError(
+                    "You cannot use the %s attribute(s) in a v1 saved query"
+                    % ", ".join(not_allowed)
+                )
+
+        if query.get("version") == 2:
+            v1_fields = set(["groupby", "rollup", "aggregations", "conditions", "limit"])
```

Comment:
It might make sense to declare these versioned fields on the class, like `{1: <fields>, 2: <fields>}`, then you can just generalize the `not_allowed` logic.